benchmarks/asdf.py

```diff
@@ -0,0 +1,82 @@
+import io
+
+import asdf
+
+from . import _utils
+
+
+def build_tree_keys():
+    return [f"{tree_size}_{data_size}" for tree_size in _utils.tree_sizes for data_size in _utils.data_sizes]
+
+
+class InitSuite:
+
+    params = build_tree_keys()
+
+    def setup(self, key):
+        self.tree = _utils.build_tree(*key.split('_'))
+
+    def time_AsdfFile_init(self, key):
+        af = asdf.AsdfFile(self.tree)
+
+    def mem_AsdfFile_init(self, key):
+        return asdf.AsdfFile(self.tree)
+
+
+class ValidateSuite:
+
+    params = build_tree_keys()
+
+    def setup(self, key):
+        self.asdf_file = asdf.AsdfFile(_utils.build_tree(*key.split('_')))
+
+    def time_AsdfFile_validate(self, key):
+        af = self.asdf_file.validate()
+
+    def peakmem_AsdfFile(self, key):
+        # peakmem includes setup, this is for comparison
+        pass
+
+    def peakmem_AsdfFile_validate(self, key):
```

Comment:
I'm not sure I fully understand `peakmem` vs `mem`. In this instance, I want to test how much memory is consumed during a call to `AsdfFile.validate`. Initially I though to use `mem` but this appears to measure memory of a specific object (which doesn't fit this use). `peakmem` seemed like a better fit but includes the memory consumed during `setup`. Next I thought to use `setup_cache` as that doesn't appear to get included with `peakmem`. However, `setup_cache` can't be parametrized. In the end I added a comparison benchmark `peakmem_AsdfFile` that does nothing and should provide a measurement of memory consumed during `setup`. Is there a better way to accomplish this? It would be nice if there was a way to snapshot the memory used and only look at a delta but I found no way to do this in asv.

Reply:
I'm not as familiar with `asv`'s memory benchmarking, but from what you said it looks like the functionality isn't quite what is needed here. We could also try `memray` or something to that effect.

Reply:
That's a great idea!

One thing I'd like to try to catch are issues where a function call (or class instantiation etc) creates objects that are difficult to garbage collect (or are erroneously cached) and create ever increasing memory usage (like https://github.com/spacetelescope/stdatamodels/pull/109). I think `memray` would be helpful for catching these.